### PR TITLE
Fix reflection to work with default interface methods

### LINKED ISSUE #15644: [Default Interfaces] Support for reflection invoke
This will currently assert in ``MemberInfoCache`1::PopulateMethods``:

```
.assembly extern mscorlib { }
.assembly simpleinvoke { }

.class interface private abstract auto ansi IFoo
{
  .method public hidebysig newslot virtual 
          instance int32  Frob() cil managed
  {
    .maxstack  8
    ldc.i4.s   100
    ret
  }
}

.class private auto ansi beforefieldinit Fooer
       extends [mscorlib]System.Object
       implements IFoo
{
  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  }
}

.method private hidebysig static int32 
        Main() cil managed
{
  .entrypoint
  .maxstack  8
  ldtoken    IFoo
  call       class [mscorlib]System.Type [mscorlib]System.Type::GetTypeFromHandle(valuetype [mscorlib]System.RuntimeTypeHandle)
  ldstr      "Frob"
  call       instance class [mscorlib]System.Reflection.MethodInfo [mscorlib]System.Type::GetMethod(string)
  newobj     instance void Fooer::.ctor()
  ldnull
  callvirt   instance object [mscorlib]System.Reflection.MethodBase::Invoke(object,
                                                                            object[])
  unbox.any  [mscorlib]System.Int32
  ret
}
```

### LINKED ISSUE #15645: [Default interfaces] Type.GetInterfaceMap support
This is a [reflection API](https://msdn.microsoft.com/en-us/library/system.type.getinterfacemap(v=vs.110).aspx) that likely needs to learn about default interfaces.

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Fixes #15645.
Fixes #15644.

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
@AtsushiKan Could you have a look please?

Cc @sergiy-k

File: src/mscorlib/src/System/RtType.cs [hunk 1]
```diff
@@ -2747,15 +2747,18 @@ public override InterfaceMapping GetInterfaceMap(Type ifaceType)
                 Debug.Assert(ifaceMethodBase is RuntimeMethodInfo);
                 im.InterfaceMethods[i] = (MethodInfo)ifaceMethodBase;
 
-                // If the slot is -1, then virtual stub dispatch is active.
-                int slot = GetTypeHandleInternal().GetInterfaceMethodImplementationSlot(ifaceRtTypeHandle, ifaceRtMethodHandle);
+                // If the impl is null, then virtual stub dispatch is active.
+                RuntimeMethodHandleInternal classRtMethodHandle = GetTypeHandleInternal().GetInterfaceMethodImplementation(ifaceRtTypeHandle, ifaceRtMethodHandle);
 
-                if (slot == -1) continue;
+                if (classRtMethodHandle.IsNullHandle()) continue;
```
Comment (AtsushiKan) [inline src/mscorlib/src/System/RtType.cs]: I know this is preexisting but the "continue" should be on its own line.

**Comment by AtsushiKan** [comment]
Other than that, LGTM

File: src/mscorlib/src/System/RtType.cs [hunk 2]
```diff
@@ -580,26 +580,25 @@ private unsafe RuntimeMethodInfo[] PopulateMethods(Filter filter)
...
                             Debug.Assert(
-                                (RuntimeMethodHandle.GetAttributes(methodHandle) & (MethodAttributes.RTSpecialName | MethodAttributes.Abstract | MethodAttributes.Virtual)) == (MethodAttributes.Abstract | MethodAttributes.Virtual) ||
-                                (RuntimeMethodHandle.GetAttributes(methodHandle) & MethodAttributes.Static) == MethodAttributes.Static ||
-                                RuntimeMethodHandle.GetName(methodHandle).Equals(".ctor") ||
-                                RuntimeMethodHandle.GetName(methodHandle).Equals(".cctor") ||
-                                RuntimeMethodHandle.GetName(methodHandle).StartsWith("IL_STUB", StringComparison.Ordinal));
+                                (RuntimeMethodHandle.GetAttributes(methodHandle) & MethodAttributes.RTSpecialName) == 0 ||
+                                RuntimeMethodHandle.GetName(methodHandle).Equals(".cctor"));
+
+                            if ((methodAttributes & MethodAttributes.RTSpecialName) != 0)
```
Comment (MichalStrehovsky) [inline src/mscorlib/src/System/RtType.cs]: I moved this check up to align with how this is done for the `else` (not interface) case below.

**Comment by MichalStrehovsky** [comment]
@AtsushiKan I had to also add what I had to do to fix reflection invoke because I didn't realize this hits that code path with the assert (I worked on several things in parallel in my local branch and Invoke didn't have tests ready yet...)

File: src/mscorlib/src/System/RtType.cs [hunk 3]
```diff
@@ -580,26 +580,25 @@ private unsafe RuntimeMethodInfo[] PopulateMethods(Filter filter)
...
+                            MethodAttributes methodAttributes = RuntimeMethodHandle.GetAttributes(methodHandle);
+
+                            #region Continue if this is a constructor
                             Debug.Assert(
-                                (RuntimeMethodHandle.GetAttributes(methodHandle) & (MethodAttributes.RTSpecialName | MethodAttributes.Abstract | MethodAttributes.Virtual)) == (MethodAttributes.Abstract | MethodAttributes.Virtual) ||
-                                (RuntimeMethodHandle.GetAttributes(methodHandle) & MethodAttributes.Static) == MethodAttributes.Static ||
-                                RuntimeMethodHandle.GetName(methodHandle).Equals(".ctor") ||
-                                RuntimeMethodHandle.GetName(methodHandle).Equals(".cctor") ||
-                                RuntimeMethodHandle.GetName(methodHandle).StartsWith("IL_STUB", StringComparison.Ordinal));
+                                (RuntimeMethodHandle.GetAttributes(methodHandle) & MethodAttributes.RTSpecialName) == 0 ||
```
Comment (AtsushiKan) [inline src/mscorlib/src/System/RtType.cs]: You could now use the cached value for methodAttributes.

**Comment by AtsushiKan** [comment]
LGTM. Is there anything that needs to happen on the Invoke path on Project N?

File: src/mscorlib/src/System/RtType.cs (hunk 3 above)
Comment (MichalStrehovsky) [reply src/mscorlib/src/System/RtType.cs]: Yeah, but I would also have to update the "not an interface" case below to do the same to stay consistent. I'm not convinced it's worth the noise in `git blame`.

**Comment by MichalStrehovsky** [comment]
> LGTM. Is there anything that needs to happen on the Invoke path on Project N?

I'll open that particular can of worms later.

Btw: do you know of any other parts of the CLR reflection stack that would need updating?

**Comment by AtsushiKan** [comment]
Reflection doesn't even expose method .overrides today so these things are going to be largely invisible to the apis that walk up base class chains to generate lists of methods. (Assuming interface defaults work the same way as other .overrides- i.e. you have to invoke them through the interface-typed `this`, not a class-based `this`) 

So I imagine any update, if any, will come in the form of requests to add or change the behavior of apis - not so much bug fixes. Or at least documentation that anticipates user confusion (why does `GetMethod()` on a class find inherited members but not `GetMethod` on an interface.)

**Pull request merged**